For User, don't query groups if the groups did not change

kbrock opened the pull request:
We were fetching all groups every time we saved.
Not such a big deal for a typical loaded user object. but for a simple
update query, this added a lot of overhead

used matching on the query count spec to avoid the save points counting towards queries

Alternative to #20471

/cc @NickLaMuro

d-m-u reviewed (approved):
lgtm 
it should be green if you kick it

kbrock commented:
rebased. hoping the changes on master fixes these test failures

d-m-u commented on spec/models/user_spec.rb:
```diff
@@ -19,6 +19,11 @@
+
+    it "doesn't validate fields that did not change. specifically group, tenant, and userid" do
+      u = FactoryBot.create(:user)
+      expect { u.update(:lastlogon => Time.zone.now) }.to make_database_queries(:count => 1, :matching => /(select|update)/i)
```
🎉

chessbyte commented:
@Fryguy @NickLaMuro Nice and simple - LGTM.  What do you think?

Fryguy reviewed (commented):
Do we have ~~treats~~ tests for a new record as well?

chessbyte commented:
@Fryguy when you say `treats`, do you really mean `treats` or `tests`?

Fryguy commented:
Damn autocorrect 😆  - yes, tests

kbrock commented:
I'll take treats

Fryguy commented:
🍰

Fryguy commented:
By new record I mean `.new`, then `.save` (or perhaps just `.create` directly).  I'd expect that we always check the groups in a new record when the current_group_id is set.

kbrock commented:
There are a bunch of tests around current group being populated and changed
added a test to ensure having a nil group is valid and does not cause an error

NickLaMuro reviewed (approved):
For testing the two PRs in question:

- https://github.com/ManageIQ/manageiq/pull/20471
- https://github.com/ManageIQ/manageiq/pull/20590

Using the following set of commands for both:

```console
$ bundle exec miqperf benchmark -ac5 "/api/users"
D, [2020-09-24T16:57:47.020314 #28653] DEBUG -- : --> logging in...
D, [2020-09-24T16:57:52.913643 #28653] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:53.197721 #28653] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:53.304659 #28653] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:53.404393 #28653] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:53.508364 #28653] DEBUG -- : --> making GET request: /api/users
$ bundle exec miqperf benchmark -ac5 "/api/users"
D, [2020-09-24T16:57:58.382058 #28728] DEBUG -- : --> logging in...
D, [2020-09-24T16:57:58.714843 #28728] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:58.813007 #28728] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:58.907880 #28728] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:59.005705 #28728] DEBUG -- : --> making GET request: /api/users
D, [2020-09-24T16:57:59.101667 #28728] DEBUG -- : --> making GET request: /api/users
$ bundle exec miqperf report --last
```


The following results show that the two approaches have equivalent effects for the API route in question:


### https://github.com/ManageIQ/manageiq/pull/20471

```
/api/auth
|   ms | queries | query (ms) | rows |
| ---: |    ---: |       ---: | ---: |
| 4878 |      20 |        197 |    8 |
|  246 |      20 |        7.3 |    8 |
/api/users
|   ms | queries | query (ms) | rows |
| ---: |    ---: |       ---: | ---: |
|  288 |      12 |         15 | 1494 |
|   18 |      10 |        3.8 |   12 |
|   16 |      10 |        3.8 |   12 |
|   17 |      10 |        3.7 |   12 |
|   19 |      10 |        4.4 |   12 |
|   16 |      10 |        2.9 |   12 |
|   16 |      10 |        3.1 |   12 |
|   16 |      10 |        3.4 |   12 |
|   14 |      10 |        2.6 |   12 |
|   21 |      10 |        4.6 |   12 |
```


### https://github.com/ManageIQ/manageiq/pull/20590

```
/api/auth
|   ms | queries | query (ms) | rows |
| ---: |    ---: |       ---: | ---: |
| 4956 |      20 |        194 |    8 |
|  247 |      20 |        7.8 |    8 |
/api/users
|   ms | queries | query (ms) | rows |
| ---: |    ---: |       ---: | ---: |
|  147 |      12 |         13 | 1494 |
|   17 |      10 |        3.5 |   12 |
|   17 |      10 |        3.5 |   12 |
|   18 |      10 |        3.8 |   12 |
|   17 |      10 |        3.8 |   12 |
|   17 |      10 |        3.1 |   12 |
|   15 |      10 |        2.8 |   12 |
|   16 |      10 |        3.2 |   12 |
|   16 |      10 |        3.1 |   12 |
|   18 |      10 |        3.2 |   12 |
```

kbrock commented:
thanks for the numbers @NickLaMuro and the once over
